### ISSUE TITLE
Track pipeline flow progress

### ISSUE THREAD

**Issue by mensfeld** [trailblazer/trailblazer#160]
Case: I have a really simple policy: if anything goes wrong - raise an exception. I know I could handle this in the endpoint but for my case it is way more sufficient to do that here. The only thing that is missing is the information on the position: which step failed which means that to obtain that, I would have to put separate failure before each step.

Could we maybe track the position in the pipe, so in case of failure we know when the track was switched? I believe it would help a lot.

```ruby
module Repositories
  class Update < Create
    contract Contracts::Create

    step Containers::Model::Find(Repository)
    step Containers::Contract::Build()
    step Containers::Contract::Validate(key: :repository)
    step Containers::Contract::Persist()

    failure Containers::Error::Raise()
  end
end
```

**Comment by apotonick** [comment]
This is planned, yes.

Why do you use exceptions? They are very hard to handle and take away flexibility you get from the pipe?

**Comment by mensfeld** [comment]
@apotonick I use exceptions because I don't want to handle them. I have a mission critical system that expects always to have a happy path and if anything goes wrong, it should raise an error.

**Comment by dnd** [comment]
What does the structure of this data look like? Is it just the name of the step that switched from right to left `results.failed_step`? 

Maybe the path that was taken as an array of arrays?

```ruby
results.path =>
[
  [:right, 'operation.new'],
  [:right, 'contract.build'],
  [:right, 'validate'],
  [:left, 'model.save'],
  [:left, 'handle_failure']
]
```

**Comment by mensfeld** [comment]
@dnd that would be awesome :) - looks great!

**Comment by apotonick** [comment]
Yeah, that looks good to me, even though we will have an unlimited number of tracks in the next version, so we might have to use something like

```ruby
[
  Flow::Left, "contract.build",
...]
```

**Comment by apotonick** [comment]
BTW, can I see some pseudo-code of how you'd use this on the outside, pleeheease?

**Comment by dnd** [comment]
Technically `Pipetree::Flow::Left` I assume? What do you mean/foresee having unlimited tracks doing to this?

```ruby
[
  [Pipetree::Flow::Left, ["contract.build", [
    [Pipetree::Flow::Right, "operation.new"],
    [Pipetree::Flow::Left, "contract.validate"]
  ]]]
]
```

This would fit it similarly with what I was proposing in the gitter chat about returning a pair, or a triple with the failed return. This could then ultimately return the value as well at the end of the failed track?

**Comment by mensfeld** [comment]
The question with multiple pipelines is: how then trbr is going to resolve that is success and what is not? 1 successful (right) and multiple left? How to switch them?

**Comment by apotonick** [comment]
@mensfeld Very simple, `Left` subclasses are always not successful and vice-versa for `Right`s, which means they are good.

For now, I would suggest we only support an invisible third left track for the `fail_early: true` option, but in later versions we could document that, once we played with that a bit more?

**Comment by dnd** [comment]
How does https://github.com/trailblazer/trailblazer-operation/issues/6 play into the multiple tracks then? It's definitely not fail early, as it is supposed to leave the operation as a success. I suppose it could also apply to that issue for a desire to immediately fail an operation and do nothing else as well. 

The distinction we need to make is, is early exit a special case, or is it a separate track(s)? That specific conversation is probably more relevant on the issue I posted above though.

**Comment by apotonick** [comment]
"Fail early" is basically a hidden additional left track. "Early success" is the pendant on the right side! :grimacing:

**Comment by aldesantis** [comment]
@mensfeld @apotonick I'm also very much interested in this, for https://github.com/pragmarb/pragma-rails/pull/13.
